stdlib/ocaml/generate.mc

```diff
@@ -0,0 +1,190 @@
+
+utest generate (let_ "abcABC/:@_'" (var_ "abcABC/:@_'"))
+with (let_ "abcABC____'" (var_ "abcABC____'")) in
+
+let testRec =
+  bind_
+    (reclets_add "abcABC/:@_'" (ulam_ "ABC123"
+                               (app_ (var_ "abcABC/:@_'")
+                                     (var_ "ABC123")))
+      reclets_empty)
+    (app_ (var_ "abcABC/:@_'") (int_ 1))
+in
+
+let testRecExpected =
+  bind_
+    (reclets_add "abcABC____'" (ulam_ "_BC123"
+                               (app_ (var_ "abcABC____'")
+                               (var_ "_BC123")))
+      reclets_empty)
+    (app_ (var_ "abcABC____'") (int_ 1))
+in
+
+utest generate testRec with testRecExpected in
+
+let mutRec =
+  bind_
+    (reclets_add "'a/b/c" (ulam_ "" (app_ (var_ "123") (var_ "")))
+      (reclets_add "123" (ulam_ "" (app_ (var_ "'a/b/c") (var_ "")))
+         reclets_empty))
+    (app_ (var_ "'a/b/c") (int_ 1))
+in
+let mutRecExpected =
+  bind_
+    (reclets_add "_a_b_c" (ulam_ "var" (app_ (var_ "_23")
+                                             (var_ "var")))
+      (reclets_add "_23" (ulam_ "var" (app_ (var_ "_a_b_c")
+                                            (var_ "var")))
+        reclets_empty))
+    (app_ (var_ "_a_b_c") (int_ 1))
+in
+
+utest generate mutRec with mutRecExpected in
+
+let parseAsMExpr = lam s.
+  use MExprParser in parseExpr (initPos "") s
+in
+
+let ocamlEval = lam p. lam strConvert.
+  let subprocess = pyimport "subprocess" in
+  let blt = pyimport "builtins" in
+  let cmd =
+        pycall blt "str"
+               (join ["print_endline (", strConvert, "(", p, "))"],)
+  in
+  let encoded = pycall cmd "encode" () in
+  let p = pycallkw subprocess "run" (["ocaml", "-stdin"],)
+                              {input=encoded, capture_output=true}
+  in
+  let stdout =
+    pycall (pycall blt "getattr" (p,"stdout")) "decode" ()
+  in
+  parseAsMExpr (pyconvert stdout)
+in
+
+let sameSemantics = lam mexprAst. lam ocamlAst.
+  let mexprVal =
+    use MExprEval in
+    eval {env = []} mexprAst
+  in
+  match mexprVal with TmConst t then
+    match t.val with CInt _ then
+      let ocamlVal = ocamlEval (expr2str ocamlAst) "string_of_int" in
+      match ocamlVal with TmConst {val = CInt _} then
+        eqExpr mexprVal ocamlVal
+      else error "Values mismatch"
+    else match t.val with CBool _ then
+      let ocamlVal = ocamlEval (expr2str ocamlAst) "string_of_bool" in
+      match ocamlVal with TmConst {val = CBool _} then
+        eqExpr mexprVal ocamlVal
+      else error "Values mismatch"
+    else match t.val with CChar _ then
+      let ocamlVal =
+        ocamlEval (expr2str ocamlAst) "Printf.sprintf \"'%c'\""
+      in
+      match ocamlVal with TmConst {val = CChar _} then
+        eqExpr mexprVal ocamlVal
+      else error "Values mismatch"
+    else error "Unsupported constant"
+  else error "Unsupported value"
+in
+
+let addInt1 = addi_ (int_ 1) (int_ 2) in
+utest addInt1 with generate (symbolize addInt1) using sameSemantics in
```

Comment:
Point worth noting based on this that will have (minor) implications on typechecking: `utest using` isn't necessarily reflexive, and further doesn't require the same type on the two things being compared.

Reply:
I guess that is a future headache, or do you propose to change something?

Reply:
Just a thing for the future